+paradise/macros

xeno-by opened the pull request:
Depends on a bunch of other pull requests opened against master. Please review those separately. Only the last three commits actually describe paradise functionality.

Currently only includes type macros. Quasiquotes and untyped macros will soon arrive.

review @paulp @adriaanm @retronym

xeno-by committed:
SI-6859 type macros no longer crash sbt

By mimicking alias types, macro types no longer trip a sanity check
in sbt 0.12 and 0.13: https://github.com/harrah/xsbt/blob/ef39aeb9c1ea443c17356695758a31152910c4a7/compile/interface/src/main/scala/xsbt/API.scala#L221.

This workaround makes me sad but it's the only option to make type macros
immediately useful. Principled fix pending.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2018/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1304/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2018/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1304/

paulp commented:
I admit not being entirely clear on what you want reviewed here and for what destination it is bound. (And the tests failed.) So I will defer any reviewing pending clarification.

xeno-by commented:
Since we don't yet have a policy for paradise branches, I suggest the following:
1) `paradise/macros` @ `scalamacros/kepler` is the integration hub for macro experiments and the source of snapshot builds. Anything can be merged there, any commit can be rewritten there. This way I can update the distribution as fast as necessary for early adopters.
2) After certain features in kepler paradise have been tested enough to suggest maturity (e.g. as it happened with introduceTopLevel), they are packed into a pull request targeting `scala/scala` - either into `master` or into `paradise/macros`, depending on whether these features require a SIP or not.
3) `paradise/macros` @ `scala/scala` is the container for reviewed code ready to be merged into `master` @ `scala/scala` pending an approval from the SIP committee (or other requirements for inclusion into Scala). It undergoes the same code review and stability procedures as `master` itself.

If this policy makes sense to you, then the answer to your questions is as follows: a) the target of this pull request is `paradise/macros`, b) please review this pull request as if it were targeting `master`. 

/cc @adriaanm @gkossakowski

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1314/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2028/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1314/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2028/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1317/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1317/

paulp commented:
Can one of the admins verify this patch?

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2031/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2031/

xeno-by commented:
The failures are genuine. Looking into them right now.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2036/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2036/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1322/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1322/

xeno-by commented:
rebased against `master`. PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1341/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1341/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2054/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2054/

xeno-by commented:
rebased against `master`. PLS REBUILD ALL

retronym commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -656,7 +659,14 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
    */
   type MacroRole = String
```
This might be safer as a value class.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -629,14 +629,56 @@ trait Namers extends MethodSynthesis {
+                enterSyntheticSym(atPos(tree.pos.makeTransparent)(emissaryDef))
+              }
+            if (vparamss.isEmpty && (mods hasFlag FINAL)) emissary setFlag FINAL
+            if (vparamss.isEmpty && (mods hasFlag OVERRIDE)) emissary setFlag OVERRIDE
```
``` scala
def copyFlag(flag: Flag) = if (vparamss.isEmpty && (mods hasFlag flag)) emissary setFlag flag
(FINAL :: OVERRIDE :: Nil) foreach copyFlag
```

retronym commented on src/reflect/scala/reflect/internal/TreeInfo.scala:
```diff
@@ -736,6 +740,18 @@ abstract class TreeInfo {
       case _ => false
     })
 
+  def repackApplyAsNew(tree: Tree): Tree = {
```
Needs a doc comment.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1364/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1364/

retronym commented on src/compiler/scala/tools/nsc/typechecker/Macros.scala:
```diff
@@ -789,6 +815,185 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
+        case Select(qual, _) => Select(qual, macroName)
+        case Ident(_) => Ident(macroName)
+      }
+      val desugared = atPos(original.pos)(gen.mkApply(macroRef, targs, argss))
```
Is it possible for `argss` to contain expanded repeated parameters (ie `xs : _*`)? If so, this `Apply` is probably wrong. See `TreeGen#mkForwarder`. If not, you could add an assertion to this effect.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -872,13 +915,23 @@ trait Namers extends MethodSynthesis {
+      val parentTrees = typer.typedParentTypes(templ)
+      val expansions = flatCollect(parentTrees) { case ExpandedIntoTemplate(templ1) => Some(templ1) }
+      expansions match {
+        case templ1 :: _ =>
```
Why is it okay to discard the tail of this list? Either add a comment to explain, or tighten up the code to enforce the `length <= 1`.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -872,13 +915,23 @@ trait Namers extends MethodSynthesis {
+        case templ1 :: _ =>
+          linkExpandeeAndExpanded(templ, templ1)
+          templateSig(templ1)
+        case _ =>
```
`case Nil =>` makes the code easier to read IMO

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2077/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2077/

retronym commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -402,6 +402,14 @@ trait Trees extends api.Trees { self: SymbolTable =>
+    // in cases when templates are involved, e.g. `class C extends B(2)` or `@annot(2)`, applications
+    // might be either term-level (call of a constructor) or type-level (type macro application)
+    // in parser, when we first encounter this ambiguity, we don't have enough information to tell those use cases apart
+    // therefore we encode such applications using Apply and figure out the rest later
```
We could traverse the AST after the typer phase to check for this. I don't think we have infrastructure for this, but I would like for each phase to contribute such pre/post-condition checks. Enforcement could be enabled/disabled with a compiler flag (in consideration of performance and unforseen false-positives)

xeno-by committed:
moves crucial part of introduceTopLevel to Run

This is done to make it easy for other parts of the compiler to use
the (quite involved) logic of introduceTopLevel.

xeno-by committed:
macro engine refactoring

Macro impl bindings now store more information in signatures.
Previously it was a flattened List[Int] corresponding to flattened paramss,
now it's List[List[Int]] to preserve the lengths of parameter lists.
Also now we distinguish between c.Expr parameters and others.

Previously actual and reference macro signatures were represented as
tuples of vparamss, rets, and sometimes tparams. Now they are all
abstracted behind MacroImplSig.

Finally this patch provides better error messages in cases of
argsc <-> paramsc and argc <-> paramc mismatches.

xeno-by committed:
type macros beta 2

No spec yet, just docs at http://docs.scala-lang.org/overviews/macros/typemacros.html.
List of known issues: https://issues.scala-lang.org/secure/IssueNavigator.jspa?mode=hide&requestId=11912.

xeno-by commented:
rebased against the latest changes in paradise. PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1381/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1381/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2093/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2093/

xeno-by commented:
Same story as with my other pull request. Will fix the failures a bit later, sorry.

xeno-by committed:
refactors macro tests

putting in a nutshell, this patch:
  * condenses some macro-XXX-a/b/c/... bundles
  * renames some tests to prepare for type macros
  * introduces some additional tests

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1384/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1384/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2096/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2096/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1393/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1393/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2122/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2122/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1412/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1412/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2136/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2136/

xeno-by commented:
```
java.util.concurrent.TimeoutException: Futures timed out after [60 seconds]
```

Dear kitty, this is so not funny. PLS REBUILD pr-scala-testsuite-linux-opt

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2141/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2141/

xeno-by commented:
I have to retract my pull requests for now - too much stuff going on. Thank you for reviews, I'll take them into account, consolidate them with my latest developments, when those stabilize, and then resubmit.
